src/components/tl/ucp/allreduce/allreduce.c

```diff
@@ -57,11 +63,121 @@ ucc_status_t ucc_tl_ucp_allreduce_knomial_init(ucc_base_coll_args_t *coll_args,
 {
-    ucc_coll_task_t *coll_task = NULL;
-    ucc_tl_ucp_allreduce_sliding_window_progress(coll_task);
-    return UCC_OK;
+    ucc_schedule_t          *schedule       = NULL;
+    ucc_status_t             status         = UCC_OK;
+    ucc_tl_ucp_team_t       *tl_team        =
+        ucc_derived_of(team, ucc_tl_ucp_team_t);
+    size_t                   allgather_size =
+        sizeof(ucc_tl_ucp_allreduce_sw_host_allgather_t);
+    ucc_rank_t               size           = UCC_TL_TEAM_SIZE(tl_team);
+    ucc_base_coll_args_t bargs    = {
+        .mask = 0,
+        .args = {
+            .coll_type = UCC_COLL_TYPE_ALLGATHER,
+            .mask      = 0,
+            .src.info = {.buffer   = NULL,
+                         .count    = allgather_size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST},
+            .dst.info = {.buffer   = NULL,
+                         .count    = allgather_size * size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST}
+        }
+    };
+    ucc_base_coll_args_t barrier_coll_args = {
+        .team           = team->params.team,
+        .args.coll_type = UCC_COLL_TYPE_BARRIER,
+    };
+    ucc_tl_ucp_allreduce_sw_host_allgather_t *allgather_data;
+    ucc_tl_ucp_task_t                        *rdma_task;
+    ucc_coll_task_t                          *barrier_task;
+
+    status = ucc_tl_ucp_get_schedule(tl_team, coll_args,
+                                    (ucc_tl_ucp_schedule_t **)&schedule);
+    if (ucc_unlikely(UCC_OK != status)) {
+        return status;
+    }
+
+    *task_h                  = &schedule->super;
+    schedule->super.post     = ucc_tl_ucp_allreduce_sliding_window_start;
+    schedule->super.progress = NULL;
+    schedule->super.finalize = ucc_tl_ucp_allreduce_sliding_window_finalize;
+
+    schedule->super.flags |= UCC_COLL_TASK_FLAG_EXECUTOR;
+
+    if (UCC_OK != status) {
+        ucc_error("failed to init executor: %s", ucc_status_string(status));
+    }
+
+    rdma_task = ucc_tl_ucp_init_task(coll_args, team);
+    if (ucc_unlikely(!rdma_task)) {
+        ucc_error("couldnt allocate task");
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    if (ucc_tl_ucp_allreduce_sliding_window_alloc_pipe(team, rdma_task) != UCC_OK) {
+        ucc_error("failed to alloc pipe: %s", ucc_status_string(status));
+        goto free_rdma_task;
+    }
+
+    status = ucc_tl_ucp_allreduce_sliding_window_task_init(coll_args, team,
+                                                           rdma_task);
+    if (status != UCC_OK) {
+        ucc_error("failed to init task: %s", ucc_status_string(status));
+        goto out;
+    }
+
+    allgather_data = rdma_task->allreduce_sliding_window.allgather_data;
+    bargs.args.src.info.buffer = allgather_data;
+    bargs.args.dst.info.buffer = PTR_OFFSET(allgather_data, allgather_size);
```

Comment:
I am using ring because knomial fails, and I measured the difference to be ~10 us. At the target message size range (over 1 MB) this difference about the same as you would see in system noise. Is it ok if I leave it?